NXP backend: Switch to the default ExecuTorch graph visualization.

MartinPavella opened the pull request:
### Summary
This PR updates the existing ExecuTorch graph visualization tool to suit NXP use-cases, and removes the old custom visualizer.

### Test plan
N/A


cc @robert-kalmar @roman-janik-nxp @StrycekSimon @jirioc

MartinPavella committed:
NXP backend: Update the ExecuTorch graph visualizer to suite NXP use-cases.

Add style .json file, which can be loaded in the Model Explorer GUI, and it changes the visualization to highlight the clusters and partitions in different colors.

MartinPavella commented:
@pytorchbot label "module: nxp" "release notes: nxp"

StrycekSimon reviewed (commented):
.

MartinPavella committed:
NXP backend: Remove old NXP graph visualization.

MartinPavella commented:
The failing linting checks are related to imports from `model_explorer`.  As far as I can tell, it is caused by missing stub files.
I checked all the [4 commits](https://github.com/pytorch/executorch/commits/main/devtools/visualization/visualization_utils.py) which modified this module before, and they all had MYPY linting errors. I don't know if I can fix it.
@Gasoonjia do you know of any way to fix this?

Erik-Lundell reviewed (commented):
Chiming in with some comments, hope that's ok.

MartinPavella commented:
> Chiming in with some comments, hope that's ok.

Thank you for the review @Erik-Lundell .
I have renamed the function and added the callbacks as you suggested. The visualization is no longer just for NXP, and it can be used by any backend.
The `model_explorer_cluster_style.json` can be loaded in the Model Explorer GUI (by clicking the palette in the top right corner), to add colors to the QDQ clusters and partitions for better readability. (I have added a note to the commit message as you suggested)

Do you have any information regarding the MYPY import warnings (the failing lintrunner)?

Erik-Lundell commented:
Thanks for your adjustments! I was probably not too clear what I meant with callback, I meant that the function takes a callback(Node) -> namespace, that makes it possible to use for any type of clustering. You can then include a new function that does the particular clustering you are interested in. But my main concerns are fixed, so I'll leave the rest of the review to someone who can actually approve. 

Regarding mypy, I don't know, sorry. Maybe just add ignore markers?

robert-kalmar reviewed (changes requested):
@MartinPavella, Implementation and concept wise looks good to me. 

What I recommend: 
* Integrate it as example usage into nxp's `aot_neutron_example` where on the CLI argument e.g. `--visualize=show|store` the example would either show the visualization of the model or store it to a json file based on the input model name. This will give everyone an example use and context for the reviewers, especially @Gasoonjia as code owner. 
* In the example store/visualize the exported quantized graph (shows the QDQ cluster visualization) and final exported program with NeutronBackend node. Optionally you can also visualize/store the output of the partitioner, although it is a bit more difficult as the CLI argument must be propagated from example to the NeutronPartitioner (e.g. via `neutron_compile_spec`, although bit hacky). Or ExecuTorch's `to_edge_transform_and_lower` and/or `to_edge` must be extended to support:
  *  either propagation of visualize argument, then the visualization can be done on the level of ExecuTorch for all the backends. Or
  *  the aforementioned routines would return the partitioned graph . 
* Document the usage in a new `md` in `docs\sources\` showing how to use the `visualize_with_cluster`, opening a model explorer, using the predefined color scheme. 
* There is no documentation available how to use the visualization at all. The `docs/source/devtools-overview.md` mentions the "visualization - Comming soon" is outdated, as the visualization is already here for some time :-). You spent time in investigation, so feel free to add short guideline how to use it, this will give you a prerequisity for the documentation of `visualize_with_cluster` and fill the gap with missing visualization documentation. 
* Move the color coding style definition (`model_explorer_cluster_style.json`) to a subfolder. This way, there will be a foundation for other backends to add their style definition to visualize their features. 

@digantdesai , for the context, at our early development days we mentioned our visualization tool, based on `.dot` files, and discussed using the ExecuTorch visualizer instead. Martin worked out a proposal. Feel free to comment on the PR or my suggestions.

roman-janik-nxp reviewed (changes requested):
Besides small adjustments and documentation it looks fine.

MartinPavella commented on devtools/visualization/visualization_utils.py:
```diff
@@ -139,6 +146,132 @@ def visualize_model_explorer(
+
+
+def visualize_with_clusters(
+    exported_program: ExportedProgram,
```
I thought it made the type annotation unnecessarily complex, even though it doesn't really provide much utility.
In the case that we have a ProgramManager, instead of using:
```python
visualize_with_cluster(edge_program_manager)
```
we have to use
```python
visualize_with_cluster(edge_program_manager.exported_program())
```
I can definitely change it to match the `visualize` function if you think that would be better. I just had a slight preference for this simpler approach.

MartinPavella committed:
NXP backend: Add documentation for graph visualization.

MartinPavella committed:
NXP backend: Add visualization support to `aot_neutron_compile.py`.

MartinPavella commented:
> @MartinPavella, Implementation and concept wise looks good to me.
> 
> What I recommend:
> 
>     * Integrate it as example usage into nxp's `aot_neutron_example` where on the CLI argument e.g. `--visualize=show|store` the example would either show the visualization of the model or store it to a json file based on the input model name. This will give everyone an example use and context for the reviewers, especially @Gasoonjia as code owner.
> 
>     * In the example store/visualize the exported quantized graph (shows the QDQ cluster visualization) and final exported program with NeutronBackend node. Optionally you can also visualize/store the output of the partitioner, although it is a bit more difficult as the CLI argument must be propagated from example to the NeutronPartitioner (e.g. via `neutron_compile_spec`, although bit hacky). Or ExecuTorch's `to_edge_transform_and_lower` and/or `to_edge` must be extended to support:
>       
>       * either propagation of visualize argument, then the visualization can be done on the level of ExecuTorch for all the backends. Or
>       * the aforementioned routines would return the partitioned graph .
> 
>     * Document the usage in a new `md` in `docs\sources\` showing how to use the `visualize_with_cluster`, opening a model explorer, using the predefined color scheme.
> 
>     * There is no documentation available how to use the visualization at all. The `docs/source/devtools-overview.md` mentions the "visualization - Comming soon" is outdated, as the visualization is already here for some time :-). You spent time in investigation, so feel free to add short guideline how to use it, this will give you a prerequisity for the documentation of `visualize_with_cluster` and fill the gap with missing visualization documentation.
> 
>     * Move the color coding style definition (`model_explorer_cluster_style.json`) to a subfolder. This way, there will be a foundation for other backends to add their style definition to visualize their features.
> 
> 
> @digantdesai , for the context, at our early development days we mentioned our visualization tool, based on `.dot` files, and discussed using the ExecuTorch visualizer instead. Martin worked out a proposal. Feel free to comment on the PR or my suggestions.

Thank you for the suggestions. I have implemented the changes except for the visualization of the partitioning. I didn't want to make such a large change to the entire pipeline within this PR.

roman-janik-nxp reviewed (approved):
Nice documentation ❤️ .

robert-kalmar commented on docs/source/devtools-overview.md:
```diff
@@ -17,7 +17,7 @@ The ExecuTorch Developer Tools support the following features:
 - **Numerical Discrepancy Detection** - Operator-level numerical discrepancy detection between AOT and runtime intermediate outputs to streamline numerical debugging and validation.
 - **Memory Allocation Insights** - Visualize how memory is planned, where all the live tensors are at any point in time
-- **Visualization** - Coming soon
+- **Visualization** - Visualize the model as a computational graph with highlighted partitions for easy debugging.
```
```suggestion
- **Visualization** - Visualize the model as a computational graph
```
Note: As we are working on it, let provide also a brief generic documentation as missing so far. Not just our contribution.

robert-kalmar commented on docs/source/visualize-with-clusters.md:
```diff
@@ -0,0 +1,78 @@
+
+### Custom Styles in `ModelExplorer`
+
+To customize the appearance of nodes:
```
There is already a color style for the QDQ cluster and partitions highlighting provided in the `devtools/visualization/model_explorer_styles/cluster_highlight_style.json`. To load it follow these steps:

mergennachin commented on docs/source/visualize.md:
```diff
@@ -0,0 +1,90 @@
+
+The function `visualize()` takes an `ExportedProgram` and launches a `ModelExplorer` server instance. A browser tab will
+open, containing the visualization.
+
```
Can you provide a screenshot pls

StrycekSimon reviewed (approved):
Have no comments here. Nice work!

MartinPavella commented on docs/source/visualize.md:
```diff
@@ -0,0 +1,144 @@
+# Visualize a Model using ModelExplorer
```
Good point. I have added the references
